src/data/nat/periodic.lean

```diff
@@ -0,0 +1,146 @@
+/-
+Copyright (c) 2021 Bolton Bailey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bolton Bailey
+-/
+import algebra.periodic
+import data.nat.interval
+import data.nat.count
+
+/-!
+# Periodic Functions on ℕ
+
+This file identifies a few functions on `ℕ` which are periodic, and also proves a lemma about
+periodic predicates which helps determine their cardinality when filtering intervals over them.
+-/
+
+namespace nat
+
+open nat function
+
+lemma periodic_gcd (a : ℕ) : periodic (gcd a) a :=
+by simp only [forall_const, gcd_add_self_right, eq_self_iff_true, periodic]
+
+lemma periodic_coprime (a : ℕ) : periodic (coprime a) a :=
+by simp only [coprime_add_self_right, forall_const, iff_self, eq_iff_iff, periodic]
+
+lemma periodic_mod (a : ℕ) : periodic (λ n, n % a) a :=
+by simp only [forall_const, eq_self_iff_true, add_mod_right, periodic]
+
+lemma _root_.function.periodic.map_mod_nat {α : Type*} {f : ℕ → α} {a : ℕ} (hf : periodic f a) :
+  ∀ n, f (n % a) = f n :=
+λ n, by { conv_rhs { rw [← nat.mod_add_div n a, mul_comm, ← nsmul_eq_mul, hf.nsmul] } }
+
+lemma mod_inj_on_Ico (n a : ℕ) : set.inj_on (λ k, k % a) (finset.Ico n (n+a)) :=
+begin
+  induction n with n ih,
+  { simp only [zero_add, nat_zero_eq_zero, Ico_zero_eq_range],
+    rintro k hk l hl (hkl : k % a = l % a),
+    simp only [finset.mem_range, finset.mem_coe] at hk hl,
+    rwa [mod_eq_of_lt hk, mod_eq_of_lt hl] at hkl, },
+  rw [Ico_succ_left_eq_erase_Ico, succ_add, Ico_succ_right_eq_insert_Ico le_self_add],
+  rintro k hk l hl (hkl : k % a = l % a),
+  have ha : 0 < a,
+  { by_contra ha, simp only [not_lt, nonpos_iff_eq_zero] at ha, simpa [ha] using hk },
+  simp only [finset.mem_coe, finset.mem_insert, finset.mem_erase] at hk hl,
+  rcases hk with ⟨hkn, (rfl|hk)⟩; rcases hl with ⟨hln, (rfl|hl)⟩,
+  { refl },
+  { rw add_mod_right at hkl,
+    suffices : l = n, { contradiction },
+    refine ih hl _ hkl.symm,
+    simp only [lt_add_iff_pos_right, set.left_mem_Ico, finset.coe_Ico, ha], },
+  { rw add_mod_right at hkl,
+    suffices : k = n, { contradiction },
+    refine ih hk _ hkl,
+    simp only [lt_add_iff_pos_right, set.left_mem_Ico, finset.coe_Ico, ha], },
+  { refine ih _ _ hkl; simp only [finset.mem_coe, hk, hl], },
+end
+
+section finset
+open finset
+
+-- what is the correct generality? It also works for `ℤ`
+lemma finset_Ico_image_mod (n a : ℕ) :
```

Comment:
This should probably move. Can you generalize it?
@YaelDillies Do you see what the correct general statement is?

Reply:
I thought about generalizing the main theorem. It's true of N, Z, and cyclic groups, but I don't know of anything else. Do we have a class for "singly-generated monoids"?
If I'm right, I think the easiest way to prove this for the other instances is just to prove it for N, then apply that as a lemma.

Reply:
Actually, even "singly-generated monoids" doesn't work, since I think Z wouldn't count as a singly-generated monoid, but rather a singly-generated group.

Reply:
Unfortunately, I think the answer is that there's no correct generality, because we don't have a "lawful mod" typeclass, only `has_mod`.

So could you put that lemma in `data.nat.interval` and provide the `int` version in `data.int.interval`?

Reply:
Lean doesn't seem to want to compile a PR with changes in both `data.nat.interval` and `data.int.interval`, so I suggest making a new PR for the int version.
I would also want to do the int version with a mod operator taking an int and a nat as input but I don't think we have that. Doesn't lean4 have better support for that kind of thing?

Reply:
Actually whatever, I'll just add `ha : 0 < a` and modify the existing proof.